plugin/evm/ftso.go

```diff
@@ -0,0 +1,410 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package evm
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/flare-foundation/flare/ids"
+
+	"github.com/flare-foundation/coreth/accounts/abi"
+	"github.com/flare-foundation/coreth/core"
+	"github.com/flare-foundation/coreth/core/vm"
+	"github.com/flare-foundation/coreth/internal/ethapi"
+)
+
+const (
+	epochsCacheSize = 16
+)
+
+type FTSO struct {
+	blockchain  *core.BlockChain
+	last        uint64
+	seconds     uint64
+	epochs      *lru.Cache
+	manager     common.Address
+	registry    common.Address
+	whitelister common.Address
+	votepower   common.Address
+	rewards     common.Address
+}
+
+func NewFTSO(blockchain *core.BlockChain) (*FTSO, error) {
+
+	epochs, _ := lru.New(epochsCacheSize)
+	f := FTSO{
+		blockchain: blockchain,
+		seconds:    0,
+		last:       0,
+		epochs:     epochs,
+	}
+
+	manager, err := f.FTSOManager()
+	if err != nil {
+		return nil, fmt.Errorf("could not get FTSO manager address: %w", err)
+	}
+
+	f.manager = manager
+
+	registry, err := f.FTSORegistry()
+	if err != nil {
+		return nil, fmt.Errorf("could not get FTSO registry: %w", err)
+	}
+
+	f.registry = registry
+
+	whitelister, err := f.VoterWhitelister()
+	if err != nil {
+		return nil, fmt.Errorf("could not get voter whitelister: %w", err)
+	}
+
+	f.whitelister = whitelister
+
+	votepower, err := f.ReadVotePowerContract()
+	if err != nil {
+		return nil, fmt.Errorf("could not get read vote power contract address: %w", err)
+	}
+
+	f.votepower = votepower
+
+	seconds, err := f.RewardEpochDurationSeconds()
+	if err != nil {
+		return nil, fmt.Errorf("could not bootstrap reward epoch duration seconds: %w", err)
+	}
+
+	f.seconds = seconds
```

Comment:
Not that it is still relevant, but that wouldn't have worked, because we needed some of these to be set, as they are needed in the functions to retrieve the other contract addresses.